src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -87,17 +77,25 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
 
             var (documentSnapshot, documentVersion) = documentInfo.Value;
 
-            var tokens = await GetSemanticTokensAsync(textDocumentIdentifier, documentSnapshot, documentVersion, range, cancellationToken);
+            // If we have a cached response, avoid computation and return early.
+            foreach (var response in _cachedResponses)
```

Comment:
With n=5 it might not matter, but this would be more efficient as as Dictionary.